src/IROperator.cpp

```diff
@@ -248,13 +273,62 @@ Expr const_false(int w) {
     return make_zero(UInt(1, w));
 }
 
+Expr lossless_cast(Type t, Expr e) {
+    if (t == e.type()) {
+        return e;
+    } else if (t.can_represent(e.type())) {
+        return cast(t, e);
+    }
+
+    if (const Cast *c = e.as<Cast>()) {
+        if (t == c->value.type()) {
```

Comment:
It's not obvious to me that that code is correct. I'll change it.

On Tue, Oct 6, 2015 at 11:41 AM, Zalman Stern <notifications@github.com>
wrote:

> In src/IROperator.cpp
> <https://github.com/halide/Halide/pull/938#discussion_r41304615>:
>
> > @@ -248,13 +273,62 @@ Expr const_false(int w) {
> >      return make_zero(UInt(1, w));
> >  }
> >
> > +Expr lossless_cast(Type t, Expr e) {
> > +    if (t == e.type()) {
> > +        return e;
> > +    } else if (t.can_represent(e.type())) {
> > +        return cast(t, e);
> > +    }
> > +
> > +    if (const Cast *c = e.as<Cast>()) {
> > +        if (t == c->value.type()) {
>
> I think a comment here explaining that widening has already been handled
> by the can_represent case would help. (I'm worried about e.g.
> cast<float>(some_double_not_exact_as_a_float) or cast<uint8_t>(1024)
> causing this code to remove the precision reducing Cast and thus returning
> the wrong (full precision) value. If I understand, that can't happen
> because these cases were handled above.)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/halide/Halide/pull/938/files#r41304615>.
>